langchain/vectorstores/redisearch.py

```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
+
+import json
+import uuid
+from typing import Any, Callable, Iterable, List, Mapping, Optional
+
+import numpy as np
+from redis.commands.search.query import Query
+
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.utils import get_from_dict_or_env
+from langchain.vectorstores.base import VectorStore
+
+
+class RediSearch(VectorStore):
+    def __init__(
+        self, redisearch_url: str, index_name: str, embedding_function: Callable
+    ):
+        """Initialize with necessary components."""
+        try:
+            import redis
+        except ImportError:
+            raise ValueError(
+                "Could not import redis python package. "
+                "Please install it with `pip install redis`."
+            )
+        self.embedding_function = embedding_function
+        self.index_name = index_name
+        try:
+            pool = redis.ConnectionPool.from_url(redisearch_url)
+            redis_client = redis.StrictRedis(connection_pool=pool)
+        except ValueError as e:
+            raise ValueError(f"Your redis connected error: {e}")
+        self.client = redis_client
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        # TODO
+        prefix = "doc"  # prefix for the document keys
+
+        ids = []
+        # Check if index exists
+        for i, text in enumerate(texts):
+            key = f"{prefix}:{uuid.uuid4().hex}"
+            metadata = metadatas[i] if metadatas else {}
+            self.client.hset(
+                key,
+                mapping={
+                    "content": text,
+                    "content_vector": np.array(
+                        self.embedding_function(text), dtype=np.float32
+                    ).tobytes(),
+                    "metadata": json.dumps(metadata),
+                },
+            )
+            ids.append(key)
+        return ids
+
+    def similarity_search(
+        self, query: str, k: int = 4, **kwargs: Any
+    ) -> List[Document]:
+        # Creates embedding vector from user query
+        embedding = self.embedding_function(query)
+
+        # Prepare the Query
+        return_fields = ["metadata", "content", "vector_score"]
+        vector_field = "content_vector"
+        hybrid_fields = "*"
```

Comment:
does the user have the option to change this here. I might be missing some logic but our users love this.

Reply:
Yea which is ok, but that's fixed for when ``metadata`` is a key. Redis allows for arbitrarily named keys to be passed in this argument. 

Assuming the user created and used the index with langchain, this is totally fine, just thinking about use cases other than the typical. again, one thing we can contribute after the initial version too.